src/goto-symex/goto_symex.cpp

```diff
@@ -87,8 +96,294 @@ void goto_symext::symex_assign(statet &state, const code_assignt &code)
+
+    // Try to constant propagate potential side effects of the assignment, when
+    // simplification is turned on and there is one thread only. Constant
+    // propagation is only sound for sequential code as here we do not take into
+    // account potential writes from other threads when propagating the values.
+    if(symex_config.simplify_opt && state.threads.size() <= 1)
+    {
+      if(constant_propagate_assignment_with_side_effects(
+           state, symex_assign, lhs, rhs))
+        return;
+    }
+
     exprt::operandst lhs_if_then_else_conditions;
-    symex_assignt{state, assignment_type, ns, symex_config, target}.assign_rec(
+    symex_assign.assign_rec(
       lhs, expr_skeletont{}, rhs, lhs_if_then_else_conditions);
   }
 }
+
+/// Maps the given array expression containing constant characters to a string
+/// containing only alphanumeric characters
+///
+/// \param char_array: array_exprt containing characters represented by
+///   expressions of kind constant_exprt and type unsignedbv_typet or
+///   signedbv_typet
+/// \return a string containing only alphanumeric characters
+static std::string get_alnum_string(const array_exprt &char_array)
+{
+  const auto &ibv_type =
+    to_integer_bitvector_type(to_array_type(char_array.type()).subtype());
+
+  const std::size_t n_bits = ibv_type.get_width();
+  CHECK_RETURN(n_bits % 8 == 0);
+
+  static_assert(CHAR_BIT == 8, "bitwidth of char assumed to be 8");
+
+  const std::size_t n_chars = n_bits / 8;
+
+  INVARIANT(
+    sizeof(std::size_t) >= n_chars,
+    "size_t shall be large enough to represent a character");
+
+  std::string result;
+
+  for(const auto &c : char_array.operands())
+  {
+    const std::size_t c_val = numeric_cast_v<std::size_t>(to_constant_expr(c));
+
+    for(std::size_t i = 0; i < n_chars; i++)
+    {
+      const char c_chunk = static_cast<char>((c_val >> (i * 8)) & 0xff);
+      result.push_back(c_chunk);
+    }
+  }
+
+  return escape_non_alnum(result);
+}
+
+bool goto_symext::constant_propagate_assignment_with_side_effects(
+  statet &state,
+  symex_assignt &symex_assign,
+  const exprt &lhs,
+  const exprt &rhs)
+{
+  if(rhs.id() == ID_function_application)
+  {
+    const function_application_exprt &f_l1 = to_function_application_expr(rhs);
+
+    if(f_l1.function().id() == ID_symbol)
+    {
+      const irep_idt &func_id =
+        to_symbol_expr(f_l1.function()).get_identifier();
+
+      if(func_id == ID_cprover_string_concat_func)
+      {
+        return constant_propagate_string_concat(state, symex_assign, f_l1);
+      }
+      else if(func_id == ID_cprover_string_empty_string_func)
+      {
+        // constant propagating the empty string always succeeds as it does
+        // not depend on potentially non-constant inputs
+        constant_propagate_empty_string(state, symex_assign, f_l1);
+        return true;
+      }
+    }
+  }
+
+  return false;
+}
+
+void goto_symext::assign_string_constant(
```

Comment:
I don't think this function belongs in the `goto_symext` class, it doesn't seem to use any of its field (apart from ns which could be given as argument). :octopus: